### PR TITLE
DataSet - Adds unit tests for add, setOptions and on/off

### PR DESCRIPTION
**Pull request by macleodbroad-wf**
@yotamberk @wimrijnders

### CODE REVIEW

File: test/DataSet.test.js [hunk 1]
```diff
@@ -1,44 +1,43 @@
...
-    data.add([
-      //{id: 3, content: 'Item 3', start: moment(now)}, // TODO: moment fails, not the same instance
-      {id: 3, content: 'Item 3', start: now},
-      {id: 4, content: 'Item 4', start: '/Date(' + now.valueOf() + ')/'}
-    ]);
+  var sort = function (a, b) {
+    return a.id > b.id;
+  };
+
+  it('can handle multiple data types', function () {
```
Comment (wimrijnders) [inline test/DataSet.test.js]: With respect to usage of variable `data`, defined globally in the `describe`-block:

You are making a potentially dangerous assumption here, namely that the tests will *all* run, and will always run in the *same order*.

From the `mocha` command line usage:

```
-g, --grep <pattern>                    only run tests matching <pattern>
-f, --fgrep <string>                    only run tests containing <string>
```

I.e. it is possible to run specific tests, not all of them. In addition, a fix to `mocha` for optionally randomizing the test order is pending.

While this is not directly an issue with the current tests, it's something which time or a user can (and therefore will) break.

My suggestions here are:

- Wrap the initialization of `data` in an initialization function, call this in every `it`-block
- *and/or* initialize the data in a `beforeEach` block.

These can be combined. The goal in any case is to allow the `it`-blocks to run independently.

File: test/DataSet.test.js [hunk 2]
```diff
@@ -1,44 +1,43 @@
...
+    }
+  });
 
-    var data = new DataSet({
-      type: {
-        start: 'Date',
-        end: 'Date'
-      }
-    });
+  // add single items with different date types
```
Comment (wimrijnders) [inline test/DataSet.test.js]: See comment gelow on initialization of used data. In any case, don't leave this init code to run 'bare'.

File: test/DataSet.test.js [hunk 3]
```diff
@@ -1,44 +1,43 @@
...
-      {id: 3, content: 'Item 3', start: now},
-      {id: 4, content: 'Item 4', start: '/Date(' + now.valueOf() + ')/'}
-    ]);
+  var sort = function (a, b) {
+    return a.id > b.id;
+  };
+
+  it('can handle multiple data types', function () {
 
     var items = data.get();
```
Comment (wimrijnders) [inline test/DataSet.test.js]: Ensure that var `data` is always in a well-defined state before  running any check on it. This comment applies to *all* usages of `data` in this file.

File: test/DataSet.test.js [hunk 4]
```diff
@@ -282,7 +285,111 @@ describe('DataSet', function () {
       {id: 3, content: 'Item 3'},
       {id: 4, content: 'Item 4'}
     ]);
+  });
+
+  describe('add', function () {
+    var dataset = new DataSet([]);
```
Comment (wimrijnders) [inline test/DataSet.test.js]: See comments in previous `describe`-blocks on usage of bare data. Please just ensure that data is re-initialized upon execution of every `it`-block.

File: test/DataSet.test.js [hunk 5]
```diff
@@ -282,7 +285,111 @@ describe('DataSet', function () {
       {id: 3, content: 'Item 3'},
       {id: 4, content: 'Item 4'}
     ]);
+  });
+
+  describe('add', function () {
```
Comment (wimrijnders) [inline test/DataSet.test.js]: 👍 For new tests. Looks clean except for the data initialization thing.
This is stuff which is assumed to 'just work'. Which is exactly the reason why they should be regression-tested.

Comment (macleodbroad-wf) [reply test/DataSet.test.js]: Has been addressed

**Comment by macleodbroad-wf** [comment]
@wimrijnders - I've addressed your comments

File: lib/DataSet.js [hunk 6]
```diff
@@ -122,7 +117,7 @@ DataSet.prototype.setOptions = function(options) {
 
 /**
  * Subscribe to an event, add an event listener
- * @param {String} event        Event name. Available events: 'put', 'update',
+ * @param {String} event        Event name. Available events: 'add', 'update',
```
Comment (wimrijnders) [inline lib/DataSet.js]: Yesterday, I learned this should be `{string}`. But I suppose that the other PR will handle that.

File: lib/DataSet.js [hunk 7]
```diff
@@ -234,6 +216,7 @@ DataSet.prototype.add = function (data, senderId) {
  * @param {Object | Array} data
  * @param {String} [senderId] Optional sender id
  * @return {Array} updatedIds     The ids of the added or updated items
+ * @throws {Error} Error - Unknown Datatype
```
Comment (wimrijnders) [inline lib/DataSet.js]: Second `Error` shouldn't be there, just the description.

File: lib/DataSet.js (hunk 6 above)
Comment (macleodbroad-wf) [reply lib/DataSet.js]: Yes, to be addressed in a follow up PR.

**Comment by yotamberk** [comment]
tests are failing again. Not because of this PR

**Pull request merged**